vowpalwabbit/loss_functions.cc

```diff
@@ -292,6 +322,58 @@ class quantileloss : public loss_function
   float tau;
 };
 
+class expectileloss : public loss_function
+{
+public:
+  expectileloss(float& q_) : q(q_) {}
+
+  std::string getType() const override { return "expectile"; }
+  float getParameter() const override { return q; }
+
+  // Expectile loss is closely related to the squared loss, but it's an assymetric function with a quantile parameter
+  // Its methods can be derived from the corresponding methods from the squared loss multiplied by the quantile value
+  float getLoss(const shared_data* sd, float prediction, float label) const override
+  {
+    float err = label - prediction;
+    return squared_loss_impl::getLoss(sd, prediction, label) * (err > 0 ? q : (1.f - q));
+  }
+
+  float getUpdate(float prediction, float label, float update_scale, float pred_per_update) const override
+  {
+    // Based on the invariant update calculation from the [Online Importance Weight Aware Updates paper](https://arxiv.org/abs/1011.1576)
+    float err = label - prediction;
+    if (err > 0) return squared_loss_impl::getUpdate(prediction, label, q * update_scale, pred_per_update);
+    return squared_loss_impl::getUpdate(prediction, label, (1.f - q) * update_scale, pred_per_update);
+  }
+
+  float getUnsafeUpdate(float prediction, float label, float update_scale) const override
+  {
+    float err = label - prediction;
+    if (err > 0) return squared_loss_impl::getUnsafeUpdate(prediction, label, q * update_scale);
+    return squared_loss_impl::getUnsafeUpdate(prediction, label, (1.f - q) * update_scale);
+  }
+
+  float first_derivative(const shared_data* sd, float prediction, float label) const override
+  {
+    float err = label - prediction;
+    return squared_loss_impl::first_derivative(sd, prediction, label) * (err > 0 ? q : (1.f - q));
+  }
+
+  float getSquareGrad(float prediction, float label) const override
```

Comment:
Move this method before first_dericative to keep the same order?

Reply:
Good point. It's fixed now.